File: Sources/agent-snap/Recorder.swift
import Foundation
import AppKit
import QuartzCore
import CoreMedia

/// Records a video of the main display plus a timestamped log of everything that happened:
/// every frame's arrival and dirty area, every gesture, every window switch. Accessibility
/// names are attached when an app answers quickly; nothing waits on them.
final class Recorder {
    let outDir: URL
    let options: Options
    /// Bundle id of the recorder UI itself; interactions with it are ignored.
    var ignoreBundleId: String?
    /// Global rect (points, top-left origin) of our own menu bar item; clicks inside are ignored.
    var ignoreRect: CGRect?
    var onStep: ((Step) -> Void)?

    private let capture = ScreenCapture()
    private var writer: VideoWriter?
    private let tap = InputTap()
    private let tracker = WindowTracker()
    private let q = DispatchQueue(label: "agent-snap.recorder")
    private let writerQ = DispatchQueue(label: "agent-snap.writer", qos: .userInitiated)
    private var ticker: DispatchSourceTimer?
    private var lastFrameHostT: Double = 0
    private lazy var coalescer = Coalescer(queue: q)

    private var session: Session
    private var t0: Double = 0
    private let frameLock = NSLock()
    private var frameLog: [FrameLog] = []
    private var cursorLog: [CursorSample] = []
    private var lastCursorT: Double = -1
    private var videoOffsetSet = false

    private struct Pending { var target: AXTarget?; var window: WindowInfo; var secure: Bool }
    private var pending: Pending?
    private var pendingIgnored = false

    private var currentWindow: WindowInfo?
    private var segmentStart: Double = 0
    private var segmentInputs = 0
    private var pollTimer: DispatchSourceTimer?
    private var stopped = false

    init(outDir: URL, options: Options = .load()) {
        self.outDir = outDir
        self.options = options
        self.session = Session(startedAt: Date(), width: 0, height: 0, scale: 1)
    }

    func start() async throws {
        try FileManager.default.createDirectory(at: outDir, withIntermediateDirectories: true)
        tracker.captureURLs = options.captureURLs
        let logURL = outDir.appendingPathComponent("ax.log")
        FileManager.default.createFile(atPath: logURL.path, contents: nil)
        if let h = try? FileHandle(forWritingTo: logURL) {
            let start = CACurrentMediaTime()
            tracker.debugLog = { line in
                h.write((String(format: "%7.2f ", CACurrentMediaTime() - start) + line + "\n").data(using: .utf8)!)
            }
        }
        t0 = CACurrentMediaTime()
        try await capture.start()
        session.width = capture.width
        session.height = capture.height
        session.scale = capture.scale
        session.video = "recording.mov"
        writer = try VideoWriter(url: outDir.appendingPathComponent("recording.mov"), width: capture.width, height: capture.height)
        writer?.log = tracker.debugLog

        capture.onFrame = { [weak self] frame, dirty in self?.handleFrame(frame, dirty) }
        coalescer.onBegin = { [weak self] raw in self?.beginGesture(raw) }
        coalescer.onGesture = { [weak self] g in self?.finishGesture(g) }
        tap.onEvent = { [weak self] raw in
            self?.q.async {
                guard let self = self, !self.stopped else { return }
                if raw.kind == .moved || raw.kind == .dragged {
                    let t = raw.t - self.t0
                    if t - self.lastCursorT >= 0.05 {
                        let p = self.capture.toPixels(raw.point)
                        self.cursorLog.append(CursorSample(t: t, x: Int(p.x), y: Int(p.y)))
                        self.lastCursorT = t
                    }
                    self.coalescer.handle(raw)
                    return
                }
                self.segmentInputs += 1
                self.coalescer.handle(raw)
            }
        }
        try tap.start()

        // Constant frame rate: when the screen is static SCK sends nothing, so re-append the last frame.
        let tick = DispatchSource.makeTimerSource(queue: writerQ)
        tick.schedule(deadline: .now(), repeating: .milliseconds(Int(1000 / VideoWriter.fps)))
        tick.setEventHandler { [weak self] in
            guard let self = self, let w = self.writer else { return }
            let now = CACurrentMediaTime()
            if now - self.lastFrameHostT >= 1.0 / Double(VideoWriter.fps) - 0.004 {
                w.appendDuplicate(at: CMClockGetTime(CMClockGetHostTimeClock()))
            }
        }
        tick.resume()
        ticker = tick

        let timer = DispatchSource.makeTimerSource(queue: q)
        timer.schedule(deadline: .now(), repeating: .milliseconds(300))
        timer.setEventHandler { [weak self] in self?.pollWindow() }
        timer.resume()
        pollTimer = timer
        fputs("recording \(capture.width)x\(capture.height) @\(capture.scale)x → \(outDir.path)\nCtrl+C to stop.\n", stderr)
    }

    func stop() async -> Session {
        pollTimer?.cancel()
        tap.stop()
        await withCheckedContinuation { (c: CheckedContinuation<Void, Never>) in
            q.async {
                self.coalescer.flushAll()
                self.stopped = true
                let now = CACurrentMediaTime() - self.t0
                if let w = self.currentWindow {
                    self.session.timeline.append(WindowSegment(start: self.segmentStart, end: now, window: w, inputEvents: self.segmentInputs))
                }
                c.resume()
            }
        }
        ticker?.cancel()
        await capture.stop()
        writerQ.sync {}
        await writer?.finish()
        session.frames = frameLock.withLock { frameLog }
        let cursor: [CursorSample] = await withCheckedContinuation { c in q.async { c.resume(returning: self.cursorLog) } }
        session.cursor = cursor
        let snapshot: Session = await withCheckedContinuation { c in q.async { c.resume(returning: self.session) } }
        let enc = JSONEncoder()
        enc.outputFormatting = [.prettyPrinted, .sortedKeys]
        enc.dateEncodingStrategy = .iso8601
        if let data = try? enc.encode(snapshot) {
            try? data.write(to: outDir.appendingPathComponent("session.json"))
        }
        fputs("video frames: \(writer?.frames ?? 0)\n", stderr)
        return snapshot
    }

    private var px: (CGRect) -> CGRect { { [capture] in capture.toPixels($0) } }

    // MARK: frames (capture queue)

    private func handleFrame(_ frame: Frame, _ dirty: [CGRect]) {
        guard let w = writer else { return }
        writerQ.async {
            w.append(frame)
            self.lastFrameHostT = frame.t
            if !self.videoOffsetSet, w.firstPTS.isValid {
                self.videoOffsetSet = true
                let off = w.firstPTS.seconds - self.t0
                self.q.async { self.session.videoOffset = off }
            }
        }
        let union = dirty.reduce(CGRect.null) { $0.union($1) }
        frameLock.lock()
        frameLog.append(FrameLog(t: frame.t - t0, dirty: union.isNull ? nil : RectI(union)))
        frameLock.unlock()
    }

    // MARK: gestures (q)

    private func beginGesture(_ raw: RawInput) {
        guard !stopped else { return }
        let win = tracker.front(toPixels: px)
        if let mine = ignoreBundleId, win.bundleId == mine { pendingIgnored = true; return }
        if let r = ignoreRect, r.contains(raw.point), raw.kind == .leftDown || raw.kind == .rightDown { pendingIgnored = true; return }
        pendingIgnored = false
        guard pending == nil else { return }
        var target: AXTarget?
        var secure = false
        switch raw.kind {
        case .leftDown, .rightDown:
            if tracker.elementAtPointIsOwnProcess(raw.point) { pendingIgnored = true; return }
            target = tracker.hitTest(raw.point, toPixels: px)
        case .keyDown:
            secure = tracker.focusedIsSecure()
            target = tracker.focused(toPixels: px)
        default: break
        }
        pending = Pending(target: target, window: win, secure: secure)
    }

    private func finishGesture(_ g: Gesture) {
        guard !stopped else { return }
        if g.kind == .cursor {
            let win = tracker.front(toPixels: px)
            if let mine = ignoreBundleId, win.bundleId == mine { return }
            let target = tracker.hitTest(g.point, toPixels: px)
            let pt = PointI(capture.toPixels(g.point))
            let secs = String(format: "%.1f", g.endT - g.startT)
            let r = Int((g.radius * capture.scale).rounded())
            let path = Int((g.pathLength * capture.scale).rounded())
            let over = target.map { " over \($0.describe)" } ?? ""
            var step = Step(index: 0, t: g.startT - t0, kind: .cursor,
                            label: "Cursor moved\(over) for \(secs)s, no click (path \(path)px, within \(r)px of (\(pt.x),\(pt.y)))",
                            point: pt, endPoint: nil, text: nil, target: target, window: win, before: nil, after: nil, dirty: nil)
            step.endT = g.endT - t0
            append(step)
            return
        }
        if pendingIgnored { pendingIgnored = false; pending = nil; return }
        guard let p = pending else { return }
        pending = nil
        let target = p.target
        let pt = PointI(capture.toPixels(g.point))
        var step = Step(index: 0, t: g.startT - t0, kind: .click, label: "", point: pt,
                        endPoint: g.endPoint.map { PointI(capture.toPixels($0)) }, text: nil, target: target,
                        window: p.window, before: nil, after: nil, dirty: nil)
        step.endT = g.endT - t0
        let tdesc = target.map { " " + $0.describe } ?? ""
        let anchorOK: Bool = {
            guard let b = target?.bounds else { return false }
            return b.w < capture.width / 2 && b.h < capture.height / 2
        }()
        switch g.kind {
        case .click: step.kind = .click; step.label = "Click\(tdesc.isEmpty ? " at (\(pt.x),\(pt.y))" : tdesc)"
        case .doubleClick: step.kind = .doubleClick; step.label = "Double-click\(tdesc)"
        case .rightClick: step.kind = .rightClick; step.label = "Right-click\(tdesc)"
        case .drag:
            step.kind = .drag
            let e = step.endPoint!
            step.label = "Drag\(tdesc) to (\(e.x),\(e.y))"
        case .type:
            step.kind = .type
            step.text = (p.secure || !options.captureTypedText) ? "[redacted]" : g.text
            step.label = "Type \"\(step.text!)\"" + (tdesc.isEmpty ? "" : " in\(tdesc)")
            if anchorOK, let b = target?.bounds { step.point = PointI(x: b.x + b.w / 2, y: b.y + b.h / 2) }
        case .key:
            step.kind = .key; step.text = g.text
            step.label = "Press \(g.text)"
            if anchorOK, let b = target?.bounds { step.point = PointI(x: b.x + b.w / 2, y: b.y + b.h / 2) }
        case .scroll:
            step.kind = .scroll
            step.label = "Scroll \(g.scrollDY < 0 ? "down" : "up")\(tdesc.isEmpty ? "" : " in\(tdesc)")"
        case .cursor:
            return
        }
        append(step)
    }

    private func append(_ step: Step) {
        var s = step
        s.index = session.steps.count + 1
        session.steps.append(s)
        fputs("[\(fmtT(s.t))] \(s.label)\n", stderr)
        onStep?(s)
    }

    // MARK: windows (q)

    private func pollWindow() {
        guard !stopped else { return }
        let w = tracker.front(toPixels: px)
        if let mine = ignoreBundleId, w.bundleId == mine { return }
        let now = CACurrentMediaTime() - t0
        if let cur = currentWindow {
            if cur.key == w.key {
                if cur != w { currentWindow = w }
                return
            }
            coalescer.flushAll()
            session.timeline.append(WindowSegment(start: segmentStart, end: now, window: cur, inputEvents: segmentInputs))
        }
        currentWindow = w
        segmentStart = now
        segmentInputs = 0
        if session.timeline.isEmpty && session.steps.isEmpty { return }
        let step = Step(index: 0, t: now, kind: .windowSwitch, label: "Switch to \(w.describe)", point: nil, endPoint: nil,
                        text: nil, target: nil, window: w, before: nil, after: nil, dirty: nil)
        append(step)
    }
}

File: crates/core/src/recorder.rs
//! Records a video of the main display plus a timestamped log of everything that happened:
//! every frame's arrival and dirty area, every gesture, every window switch. Accessibility
//! names are attached when an app answers quickly; nothing waits on them.
//!
//! Threads: the backend delivers frames and input on its own threads; a 30 fps ticker thread feeds
//! the encoder (re-sending the last frame when nothing changed); a worker thread owns the window
//! tracker, the coalescer and the session and does all bookkeeping.

use std::path::{Path, PathBuf};
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::mpsc::{self, RecvTimeoutError, Sender};
use std::sync::{Arc, Mutex};
use std::thread::JoinHandle;
use std::time::{Duration, Instant};

use anyhow::{Context, Result};

use crate::encoder::{VideoWriter, FPS};
use crate::model::{fmt_t, CursorSample, FrameLog, PointI, RectI, Session, Step, StepKind, WindowSegment};
use crate::options::Options;
use crate::platform::{AxTarget, Backend, CaptureInfo, Frame, RawInput, RawKind, ScreenCapture, WindowInfo, WindowTracker};
use crate::semantics::{CoalesceEvent, Coalescer, Gesture, GestureKind};

pub const VIDEO_FILE: &str = "recording.mp4";
const POLL_INTERVAL: Duration = Duration::from_millis(300);
const CURSOR_SAMPLE_GAP: f64 = 0.05;

type StepCallback = Arc<dyn Fn(&Step) + Send + Sync>;

enum Msg {
    Input(RawInput),
    Stop,
}

#[derive(Default)]
struct Shared {
    /// Newest captured frame and whether it has not been written yet.
    latest: Mutex<(Option<Frame>, bool)>,
    frame_log: Mutex<Vec<FrameLog>>,
    video_offset: Mutex<Option<f64>>,
    stop: AtomicBool,
}

struct Running {
    tx: Sender<Msg>,
    worker: JoinHandle<(Session, Vec<CursorSample>)>,
    ticker: JoinHandle<VideoWriter>,
    capture: Box<dyn ScreenCapture>,
    input: Box<dyn crate::platform::InputTap>,
    shared: Arc<Shared>,
}

pub struct Recorder {
    out_dir: PathBuf,
    options: Options,
    /// Bundle id / process id of the recorder UI itself; interactions with it are ignored.
    pub ignore_bundle_id: Option<String>,
    /// Global rect (points, top-left origin) of our own tray/menu item; clicks inside are ignored.
    pub ignore_rect: Option<RectI>,
    on_step: Option<StepCallback>,
    backend: Option<Backend>,
    running: Option<Running>,
}

impl Recorder {
    pub fn new(backend: Backend, out_dir: impl Into<PathBuf>, options: Options) -> Self {
        Recorder {
            out_dir: out_dir.into(),
            options,
            ignore_bundle_id: None,
            ignore_rect: None,
            on_step: None,
            backend: Some(backend),
            running: None,
        }
    }

    pub fn out_dir(&self) -> &Path {
        &self.out_dir
    }

    /// Called on the worker thread for every appended step.
    pub fn on_step(&mut self, f: impl Fn(&Step) + Send + Sync + 'static) {
        self.on_step = Some(Arc::new(f));
    }

    pub fn is_running(&self) -> bool {
        self.running.is_some()
    }

    pub fn start(&mut self) -> Result<()> {
        if self.running.is_some() {
            anyhow::bail!("already recording");
        }
        let mut backend = self.backend.take().context("recorder already used")?;
        std::fs::create_dir_all(&self.out_dir).with_context(|| format!("creating {}", self.out_dir.display()))?;
        let t0 = Instant::now();
        let shared = Arc::new(Shared::default());

        let sh = shared.clone();
        let info: CaptureInfo = backend.capture.start(
            FPS,
            Box::new(move |frame: Frame| {
                let mut l = sh.latest.lock().unwrap_or_else(|e| e.into_inner());
                *l = (Some(frame), true);
            }),
        )?;

        let mut session = Session::new(info.width, info.height, info.scale);
        session.video = Some(VIDEO_FILE.to_string());
        let writer = VideoWriter::new(&self.out_dir.join(VIDEO_FILE), info.width, info.height)?;
        let encoder = writer.encoder().to_string();

        let ticker = {
            let sh = shared.clone();
            let (w, h) = (info.width, info.height);
            std::thread::Builder::new()
                .name("agent-snap.ticker".into())
                .spawn(move || tick_loop(writer, sh, t0, w, h))
                .context("spawning ticker")?
        };

        let (tx, rx) = mpsc::channel::<Msg>();
        let worker = {
            let mut w = Worker {
                tracker: backend.windows,
                coalescer: Coalescer::new(),
                session,
                info: info.clone(),
                options: self.options.clone(),
                t0,
                ignore_bundle_id: self.ignore_bundle_id.clone(),
                ignore_rect: self.ignore_rect,
                on_step: self.on_step.clone(),
                cursor_log: Vec::new(),
                last_cursor_t: -1.0,
                pending: None,
                pending_ignored: false,
                current_window: None,
                segment_start: 0.0,
                segment_inputs: 0,
                stopped: false,
            };
            std::thread::Builder::new()
                .name("agent-snap.recorder".into())
                .spawn(move || {
                    w.run(rx);
                    (w.session, w.cursor_log)
                })
                .context("spawning worker")?
        };

        let itx = tx.clone();
        backend.input.start(Box::new(move |raw: RawInput| {
            let _ = itx.send(Msg::Input(raw));
        }))?;

        log::info!("recording {}x{} @{}x ({encoder}) -> {}", info.width, info.height, info.scale, self.out_dir.display());
        self.running = Some(Running { tx, worker, ticker, capture: backend.capture, input: backend.input, shared });
        Ok(())
    }

    /// Stops everything, writes `session.json` and returns the session.
    pub fn stop(&mut self) -> Result<Session> {
        let mut r = self.running.take().context("not recording")?;
        r.input.stop();
        let _ = r.tx.send(Msg::Stop);
        let (mut session, cursor) = r.worker.join().map_err(|_| anyhow::anyhow!("recorder worker panicked"))?;
        r.shared.stop.store(true, Ordering::SeqCst);
        let writer = r.ticker.join().map_err(|_| anyhow::anyhow!("ticker panicked"))?;
        r.capture.stop();
        let frames = writer.frames();
        if let Err(e) = writer.finish() {
            log::error!("video writer: {e:#}");
        }
        session.frames = std::mem::take(&mut *r.shared.frame_log.lock().unwrap_or_else(|e| e.into_inner()));
        session.cursor = cursor;
        session.video_offset = r.shared.video_offset.lock().unwrap_or_else(|e| e.into_inner()).unwrap_or(0.0);
        if frames == 0 {
            session.video = None;
        }
        session.save(&self.out_dir)?;
        log::info!("video frames: {frames}");
        Ok(session)
    }
}

/// Constant frame rate: one frame per tick, re-sending the last one when the screen was static.
fn tick_loop(mut writer: VideoWriter, sh: Arc<Shared>, t0: Instant, w: u32, h: u32) -> VideoWriter {
    let period = Duration::from_secs_f64(1.0 / FPS as f64);
    let mut first: Option<Instant> = None;
    let mut n: u64 = 0;
    'outer: loop {
        if sh.stop.load(Ordering::Relaxed) {
            break;
        }
        let nominal = match first {
            Some(f) => {
                let target = f + period * n as u32;
                while Instant::now() < target {
                    if sh.stop.load(Ordering::Relaxed) {
                        break 'outer;
                    }
                    std::thread::sleep((target - Instant::now()).min(Duration::from_millis(2)));
                }
                target
            }
            None => {
                let has = sh.latest.lock().unwrap_or_else(|e| e.into_inner()).0.is_some();
                if !has {
                    std::thread::sleep(Duration::from_millis(5));
                    continue;
                }
                let now = Instant::now();
                first = Some(now);
                *sh.video_offset.lock().unwrap_or_else(|e| e.into_inner()) = Some(now.duration_since(t0).as_secs_f64());
                now
            }
        };
        let (frame, is_new) = {
            let mut l = sh.latest.lock().unwrap_or_else(|e| e.into_inner());
            let out = (l.0.clone(), l.1);
            l.1 = false;
            out
        };
        let Some(frame) = frame else { continue };
        let ok = if is_new { writer.write(&frame) } else { writer.write_duplicate() };
        if ok {
            let dirty = if is_new {
                Some(frame.dirty.iter().fold(None, |acc: Option<RectI>, d| Some(acc.map_or(*d, |a| a.union(d)))).unwrap_or(RectI::new(0, 0, w as i32, h as i32)))
            } else {
                None
            };
            sh.frame_log.lock().unwrap_or_else(|e| e.into_inner()).push(FrameLog { t: nominal.duration_since(t0).as_secs_f64(), dirty });
        }
        n += 1;
    }
    writer
}

struct Pending {
    target: Option<AxTarget>,
    window: WindowInfo,
    secure: bool,
}

struct Worker {
    tracker: Box<dyn WindowTracker>,
    coalescer: Coalescer,
    session: Session,
    info: CaptureInfo,
    options: Options,
    t0: Instant,
    ignore_bundle_id: Option<String>,
    ignore_rect: Option<RectI>,
    on_step: Option<StepCallback>,
    cursor_log: Vec<CursorSample>,
    last_cursor_t: f64,
    pending: Option<Pending>,
    pending_ignored: bool,
    current_window: Option<WindowInfo>,
    segment_start: f64,
    segment_inputs: usize,
    stopped: bool,
}

impl Worker {
    fn run(&mut self, rx: mpsc::Receiver<Msg>) {
        let mut next_poll = Instant::now();
        loop {
            let now = Instant::now();
            let mut deadline = next_poll;
            if let Some(d) = self.coalescer.next_deadline() {
                deadline = deadline.min(d);
            }
            let wait = deadline.saturating_duration_since(now);
            match rx.recv_timeout(wait) {
                Ok(Msg::Input(raw)) => self.handle_input(raw),
                Ok(Msg::Stop) | Err(RecvTimeoutError::Disconnected) => break,
                Err(RecvTimeoutError::Timeout) => {}
            }
            let now = Instant::now();
            self.coalescer.poll(now);
            self.process_events();
            if now >= next_poll {
                self.poll_window();
                next_poll = now + POLL_INTERVAL;
            }
        }
        self.coalescer.flush_all();
        self.process_events();
        self.stopped = true;
        let now = self.elapsed_now();
        if let Some(w) = self.current_window.take() {
            self.session.timeline.push(WindowSegment { start: self.segment_start, end: now, window: w, input_events: self.segment_inputs });
        }
    }

    fn elapsed(&self, t: Instant) -> f64 {
        t.saturating_duration_since(self.t0).as_secs_f64()
    }
    fn elapsed_now(&self) -> f64 {
        self.elapsed(Instant::now())
    }
    fn px(&self, x: f64, y: f64) -> PointI {
        PointI::from_f64(x * self.info.scale, y * self.info.scale)
    }
    fn px_rect(&self, r: &RectI) -> RectI {
        let s = self.info.scale;
        RectI::from_f64(r.x as f64 * s, r.y as f64 * s, r.w as f64 * s, r.h as f64 * s)
    }
    fn px_target(&self, mut t: AxTarget) -> AxTarget {
        t.bounds = t.bounds.map(|b| self.px_rect(&b));
        t
    }

    /// Frontmost window with URL (when enabled), locality flag and pixel bounds.
    fn front_window(&mut self) -> WindowInfo {
        let mut w = self.tracker.front();
        if !self.options.capture_urls {
            w.url = None;
        } else if w.url.is_none() {
            w.url = self.tracker.tab_url(&w);
        }
        w.is_local = w.url.as_deref().map(is_local_url);
        w.bounds = self.px_rect(&w.bounds);
        w
    }

    fn handle_input(&mut self, raw: RawInput) {
        if self.stopped {
            return;
        }
        if matches!(raw.kind, RawKind::Move | RawKind::Drag(_)) {
            let t = self.elapsed(raw.t);
            if t - self.last_cursor_t >= CURSOR_SAMPLE_GAP {
                let p = self.px(raw.x, raw.y);
                self.cursor_log.push(CursorSample { t, x: p.x, y: p.y });
                self.last_cursor_t = t;
            }
            self.coalescer.handle(raw);
            self.process_events();
            return;
        }
        self.segment_inputs += 1;
        self.coalescer.handle(raw);
        self.process_events();
    }

    fn process_events(&mut self) {
        for ev in self.coalescer.drain() {
            match ev {
                CoalesceEvent::Begin(raw) => self.begin_gesture(&raw),
                CoalesceEvent::Gesture(g) => self.finish_gesture(&g),
            }
        }
    }

    fn is_mine(&self, w: &WindowInfo) -> bool {
        self.ignore_bundle_id.as_deref().is_some_and(|m| !m.is_empty() && m == w.bundle_id)
    }

    fn begin_gesture(&mut self, raw: &RawInput) {
        if self.stopped {
            return;
        }
        let win = self.front_window();
        if self.is_mine(&win) {
            self.pending_ignored = true;
            return;
        }
        if let Some(r) = self.ignore_rect {
            if matches!(raw.kind, RawKind::Down(_)) && r.contains_point(raw.x, raw.y) {
                self.pending_ignored = true;
                return;
            }
        }
        self.pending_ignored = false;
        if self.pending.is_some() {
            return;
        }
        let mut target = None;
        let mut secure = false;
        match raw.kind {
            RawKind::Down(_) => {
                target = self.tracker.hit_test(raw.x, raw.y).map(|t| self.px_target(t));
            }
            RawKind::KeyDown { .. } => {
                secure = self.tracker.focused_is_secure();
                target = self.tracker.focused().map(|t| self.px_target(t));
            }
            _ => {}
        }
        self.pending = Some(Pending { target, window: win, secure });
    }

    fn finish_gesture(&mut self, g: &Gesture) {
        if self.stopped {
            return;
        }
        let start = self.elapsed(g.start_t);
        let end = self.elapsed(g.end_t);
        if g.kind == GestureKind::Cursor {
            let win = self.front_window();
            if self.is_mine(&win) {
                return;
            }
            let target = self.tracker.hit_test(g.x, g.y).map(|t| self.px_target(t));
            let pt = self.px(g.x, g.y);
            let secs = format!("{:.1}", end - start);
            let r = (g.radius * self.info.scale).round() as i64;
            let path = (g.path_length * self.info.scale).round() as i64;
            let over = target.as_ref().map(|t| format!(" over {}", t.describe())).unwrap_or_default();
            let label = format!("Cursor moved{over} for {secs}s, no click (path {path}px, within {r}px of ({},{}))", pt.x, pt.y);
            let mut step = Step::new(start, StepKind::Cursor, label, win);
            step.point = Some(pt);
            step.target = target;
            step.end_t = Some(end);
            self.append(step);
            return;
        }
        if self.pending_ignored {
            self.pending_ignored = false;
            self.pending = None;
            return;
        }
        let Some(p) = self.pending.take() else { return };
        let target = p.target;
        let pt = self.px(g.x, g.y);
        let mut step = Step::new(start, StepKind::Click, "", p.window);
        step.point = Some(pt);
        step.end_point = g.end_point.map(|(x, y)| self.px(x, y));
        step.end_t = Some(end);
        let tdesc = target.as_ref().map(|t| format!(" {}", t.describe())).unwrap_or_default();
        let anchor_ok = target
            .as_ref()
            .and_then(|t| t.bounds)
            .is_some_and(|b| b.w < self.info.width as i32 / 2 && b.h < self.info.height as i32 / 2);
        let anchor = target.as_ref().and_then(|t| t.bounds).map(|b| PointI::new(b.x + b.w / 2, b.y + b.h / 2));
        match g.kind {
            GestureKind::Click => {
                step.kind = StepKind::Click;
                step.label = if tdesc.is_empty() { format!("Click at ({},{})", pt.x, pt.y) } else { format!("Click{tdesc}") };
            }
            GestureKind::DoubleClick => {
                step.kind = StepKind::DoubleClick;
                step.label = format!("Double-click{tdesc}");
            }
            GestureKind::RightClick => {
                step.kind = StepKind::RightClick;
                step.label = format!("Right-click{tdesc}");
            }
            GestureKind::Drag => {
                step.kind = StepKind::Drag;
                let e = step.end_point.unwrap_or(pt);
                step.label = format!("Drag{tdesc} to ({},{})", e.x, e.y);
            }
            GestureKind::Type => {
                step.kind = StepKind::Type;
                let text = if p.secure || !self.options.capture_typed_text { "[redacted]".to_string() } else { g.text.clone() };
                step.label = format!("Type \"{text}\"{}", if tdesc.is_empty() { String::new() } else { format!(" in{tdesc}") });
                step.text = Some(text);
                if anchor_ok {
                    step.point = anchor;
                }
            }
            GestureKind::Key => {
                step.kind = StepKind::Key;
                step.text = Some(g.text.clone());
                step.label = format!("Press {}", g.text);
                if anchor_ok {
                    step.point = anchor;
                }
            }
            GestureKind::Scroll => {
                step.kind = StepKind::Scroll;
                let dir = if g.scroll_dy > 0.0 { "down" } else { "up" };
                step.label = format!("Scroll {dir}{}", if tdesc.is_empty() { String::new() } else { format!(" in{tdesc}") });
            }
            GestureKind::Cursor => return,
        }
        step.target = target;
        self.append(step);
    }

    fn append(&mut self, mut step: Step) {
        step.index = self.session.steps.len() + 1;
        log::info!("[{}] {}", fmt_t(step.t), step.label);
        if let Some(cb) = &self.on_step {
            cb(&step);
        }
        self.session.steps.push(step);
    }

    fn poll_window(&mut self) {
        if self.stopped {
            return;
        }
        let w = self.front_window();
        if self.is_mine(&w) {
            return;
        }
        let now = self.elapsed_now();
        if let Some(cur) = self.current_window.clone() {
            if cur.key() == w.key() {
                if cur != w {
                    self.current_window = Some(w);
                }
                return;
            }
            self.coalescer.flush_all();
            self.process_events();
            self.session.timeline.push(WindowSegment { start: self.segment_start, end: now, window: cur, input_events: self.segment_inputs });
        }
        self.current_window = Some(w.clone());
        self.segment_start = now;
        self.segment_inputs = 0;
        if self.session.timeline.is_empty() && self.session.steps.is_empty() {
            return;
        }
        let step = Step::new(now, StepKind::WindowSwitch, format!("Switch to {}", w.describe()), w);
        self.append(step);
    }
}

/// True for localhost-style hosts (same rule as the Swift tracker).
pub fn is_local_url(url: &str) -> bool {
    let rest = url.split_once("://").map(|(_, r)| r).unwrap_or(url);
    let hostport = rest.split(['/', '?', '#']).next().unwrap_or("");
    let hostport = hostport.rsplit('@').next().unwrap_or(hostport);
    let host = if hostport.starts_with('[') {
        hostport.trim_start_matches('[').split(']').next().unwrap_or("")
    } else {
        hostport.split(':').next().unwrap_or("")
    };
    let host = host.to_ascii_lowercase();
    host == "localhost" || host == "127.0.0.1" || host == "0.0.0.0" || host.ends_with(".local") || host.ends_with(".localhost")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn local_urls() {
        assert!(is_local_url("http://localhost:3000/x"));
        assert!(is_local_url("http://127.0.0.1/"));
        assert!(is_local_url("https://app.localhost/a?b=c"));
        assert!(is_local_url("http://mac.local:8080"));
        assert!(!is_local_url("https://example.com/localhost"));
        assert!(!is_local_url("https://localhost.example.com"));
    }
}
